datafusion/common/src/dfschema.rs

```diff
@@ -102,8 +217,12 @@ impl DFSchema {
                 ));
             }
         }
+
+        let fields_index = build_index(&fields);
```

Comment:
If the index is built for all DFSchema that are created, I wonder if that will be too much overhead. Maybe we could consider creating it on first use 🤔  or finding some way to canonicalize / cache the map

Reply:
this use case might indeed be a good call of interior mutability, i.e. use an RWLock and init the lookup table on the first use

Reply:
Thinking about this more: instead of `RwLock`, this can be solved even more elegantly w/ [`OnceLock::get_or_init`](https://doc.rust-lang.org/std/sync/struct.OnceLock.html#method.get_or_init) (this is usually used for static variables, but you can totally use that for struct members as well).